src/Persistence/MassTransit.Azure.Storage/AzureStorage/MessageData/AzureStorageMessageDataRepository.cs

```diff
@@ -128,7 +132,23 @@ public async Task<Uri> Put(Stream stream, TimeSpan? timeToLive = default, Cancel
             var blobName = _nameGenerator.GenerateBlobName();
             var blob = _container.GetBlobClient(blobName);
 
-            await blob.UploadAsync(stream, cancellationToken).ConfigureAwait(false);
+            if (_compress)
+            {
+                blobName += ".gz";
+                blob = _container.GetBlobClient(blobName);
```

Comment:
Determine the blob name before calling GetBlobClient, and only call it once (the call for the original filename is above).